binaries/runtime/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+#![warn(unsafe_op_in_unsafe_fn)]
+
+use dora_core::{
+    config::{CommunicationConfig, DataId, NodeId, OperatorId},
+    descriptor::OperatorDefinition,
+};
+use dora_node_api::{
+    self,
+    communication::{self, CommunicationLayer, Publisher, STOP_TOPIC},
+};
+use eyre::{bail, Context};
+use futures::{Stream, StreamExt};
+use operator::{spawn_operator, OperatorEvent};
+
+use std::{
+    collections::{BTreeSet, HashMap},
+    mem,
+};
+use tokio::{runtime::Builder, sync::mpsc};
+use tokio_stream::{wrappers::ReceiverStream, StreamMap};
+
+mod operator;
+
+pub fn main() -> eyre::Result<()> {
+    set_up_tracing().context("failed to set up tracing subscriber")?;
+
+    let node_id: NodeId = {
+        let raw =
+            std::env::var("DORA_NODE_ID").wrap_err("env variable DORA_NODE_ID must be set")?;
+        serde_yaml::from_str(&raw).context("failed to deserialize operator config")?
+    };
+    let communication_config: CommunicationConfig = {
+        let raw = std::env::var("DORA_COMMUNICATION_CONFIG")
+            .wrap_err("env variable DORA_COMMUNICATION_CONFIG must be set")?;
+        serde_yaml::from_str(&raw).context("failed to deserialize communication config")?
+    };
+    let operators: Vec<OperatorDefinition> = {
+        let raw =
+            std::env::var("DORA_OPERATORS").wrap_err("env variable DORA_OPERATORS must be set")?;
+        serde_yaml::from_str(&raw).context("failed to deserialize operator config")?
+    };
+
+    let mut communication: Box<dyn CommunicationLayer> =
+        communication::init(&communication_config)?;
+
+    let mut operator_events = StreamMap::new();
+    let mut operator_stop_publishers = HashMap::new();
+    let mut operator_events_tx = HashMap::new();
+
+    for operator_config in &operators {
+        let (events_tx, events) = mpsc::channel(1);
+        let stop_publisher = publisher(
+            &node_id,
+            operator_config.id.clone(),
+            STOP_TOPIC.to_owned().into(),
+            communication.as_mut(),
+        )
+        .with_context(|| {
+            format!(
+                "failed to create stop publisher for operator {}",
+                operator_config.id
+            )
+        })?;
+        operator_stop_publishers.insert(operator_config.id.clone(), stop_publisher);
+
+        operator_events.insert(operator_config.id.clone(), ReceiverStream::new(events));
+        operator_events_tx.insert(operator_config.id.clone(), events_tx);
+    }
+
+    let operator_events = operator_events.map(|(id, event)| Event::Operator { id, event });
+    let node_id_clone = node_id.clone();
+    let stop_thread = std::thread::spawn(move || {
+        Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(run(
+                node_id_clone,
+                operator_events,
+                operator_stop_publishers,
+            ))
+            .unwrap();
```

Comment:
I don't like the `unwrap` calls here. Errors in the `run` function are turned into panics this way, which are much more difficult to deal with. Also, we don't know about any errors until we call `stop_thread.join()` in the end, which can lead to deadlocks when an operator waits for a `stopped` message that won't be sent anymore (as the `run` loop is no longer running).

Reply:
Ok, I'll remove them

Reply:
The race condition you're mentioning seems to be unavoidable with Python and the `GIL`. Avoiding it would require spawning the python operator in another thread which is impossible.

Is it ok to leave it as is for now?

Reply:
I guess it's ok for now, given that we're mostly working with runtimes that run a single Python operator.